agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/localstorage/LocalFileLoader.java

```diff
@@ -86,20 +93,41 @@ ByteBuffer loadTelemetriesFromDisk() {
       return null;
     }
 
-    try {
-      // TODO (heya) backoff and retry delete when it fails.
-      Files.delete(tempFile.toPath());
-    } catch (IOException ex) {
-      // TODO (heya) track deserialization failure via Statsbeat
-      operationLogger.recordFailure("Fail to read telemetry from " + tempFile.getName(), ex);
-      return null;
-    } catch (SecurityException ex) {
-      operationLogger.recordFailure(
-          "Unable to delete " + tempFile.getName() + ". Access is denied.", ex);
-      return null;
-    }
-
+    toBeDeletedFileQueue.add(
+        sourceFile); // mark source file to be deleted when it's sent successfully.
+    deleteFile(tempFile); // delete temp file immediately
     operationLogger.recordSuccess();
     return ByteBuffer.wrap(result);
   }
+
+  // either delete it permanently on success or add it back to cache to be processed again later on
+  // failure
+  public void updateProcessedFileStatus(boolean success) {
+    if (success) {
+      deleteFilePermanentlyOnSuccess();
+    } else {
+      // add the file back to local file cache to be process later.
+      localFileCache.addPersistedFilenameToMap(toBeDeletedFileQueue.poll().getName());
+    }
+  }
+
+  // delete a file on the queue permanently when http response returns success.
+  private void deleteFilePermanentlyOnSuccess() {
+    File file = toBeDeletedFileQueue.poll();
```

Comment:
the queue is not clear to me, how do we know the head is the one we want to delete?

Reply:
after we load a persisted file from the disk, we sent it to breeze right away. base on the http response, if it's successful, we delete it right away; otherwise add it back to the localfilecache.  during this whole process, there will be one file on the tobedeletedQueue.

Reply:
LocalFileSender is a single thread scheduler.